\chapter{Introduction} \label{chap:1}

Virus research is a crucial field of molecular biological science, especially in the light of new pandemics nowadays. While not harming their natural host with any sign of illness%\DD{ich wei√ü was du sagen willst, aber "in any way" ist denke ich nicht so passend, eher "without any signs of illness" oder so}
, the %\DD{the?} 
viral infections, may lead to death or cause tremendous late effects in an accidental one \autocite{accidential}. In many cases, humans can be these accidental hosts. %\DD{Fehlwirt = "accidental host" oder "dead end host"}
An accidental %\DD{hier auch, oder hast du das ausm paper?} 
host is a organism with permissive cells which allow the virus to replicate by circumvent the native defenses but, unlike the natural host, with pathogenic consequences. In contrast, the infection of a natural host leads to a lifelong infection, with high viral loads but no negative effects on the organism \autocite{accidential}. Becoming the accidental host of a virus can happen by close contact with an already infected natural or accidental host, e.g. a human who has close contact to a mallard duck infected by the \gls{IAV} \autocite{duck_natural}. The \gls{IAV} is the most widespread species of the virus family \textit{Orthomyxoviridae} and also commonly known as flu \autocite{influenza_all}. While the mallard duck is one of the natural host of the \gls{IAV} and the infection shows therefore no pathogenic behavior, the subsequent infection of the human shows a high immune response related with fever and a possible mortal outcome \autocite{infection}. The process of a human to get infected by a pathogen from a non-human animal is called zoonosis. Known examples are avian flu, swine flu and HIV \autocite{avian_swine, hiv}.

By entering the organism in a way that allows the virus particles to access types of cells that provide an opportunity to replicate themselves, an %\DD{an} 
organism can become infected. The \textit{Orthomyxoviridae} can only replicate themselves in epithelial respiratory cells, so the particles need to enter tissue of the respiratory tract, e.g. the lungs \autocite{infection}. The complete virus particles are called virions. They contain all necessary informations to build new virions, by replication of its genome and arrangement of new build proteins around it, to secure and enable the transport of the genome. The structure around the genome is called capsid \autocite{buch}. Outside of a host cell, viral genomes are highly fragile and can be easily destroyed by physical forces or proteins build for degradation of nucleic acids like nucleases \autocite{buch}. Capsids are most commonly formed as icosahedrons which is a highly effective trade-off between stability, volume and surface area. While protecting the genome from the hostile environment around the cells, they enable the injection of the genome into the cell by proteins on their surface \autocite{buch}. After replication of the virions inside the host cell and the following cell death, the new particles can infect other host cells and repeat the process.

\textit{Orthomyxoviridae} genomes consist of negative \gls{ssRNA}, that is not coated by a capsid. Instead the genome is surrounded by three layers of protection. The first layer consists of \glspl{NP} bound to the genome \autocite{influenza_all}. The \glspl{NP} are surrounded by proteins called \gls{M1} building the matrix as second layer and coated by the lipid envelope build from host cell lipids, as third and final layer \autocite{influenza_all}. Viruses possessing this more complex composition are called enveloped viruses. These three layers allow a more effective production of new virions because the host cell do not % \DD{entweder "cells do not" oder "cell does not" und auch "'t" und andere Abkuerzungen in Wissenschaftlichen Arbeiten vermeiden} 
need to die to release new virions and can be slaved as a manufactory instead \autocite{buch}. To enter the host cell for replication, a the protein on the surface called \gls{HA} attaches the virion to the sialic acid receptor \autocite{influenza_all}. The complete virion is then taken into the cell by a process called endocytose. After entering the cell, the lipid envelope is fused with the vesicle membrane of the host cell to migrate the genome into the nucleus for replication and production of proteins \autocite{influenza_all}. The genome is replicated in the host nucleus by the RNA polymerase complex, already bound to the virus genome consisting of \gls{PA}, \gls{PB1} and \gls{PB2} \autocite{inf_natural}. Simultaneously, the genome is transcribed to positive stranded \gls{mRNA} by the same polymerase complex and translated into proteins needed for the same process in the new virion. Export of the replicated genome from the nucleus is induced by the \gls{NEP} \autocite{influenza_all}. After assembly of the build proteins and replicated genome into a new virion, it leaves the cell through the plasma membrane with the help of \gls{NA} while coating itself in its lipids \autocite{influenza_all}. All new virions created from the infection by one specific strain are called quasispecies \autocite{quasispecies}.

In the virions lifecycle, synthesizing of proteins is inevitable. Protection of the new replicated genome by \glspl{NP}, the building of its replication machinery from \gls{PA}, \gls{PB1} and \gls{PB2}, enable nuclear transport by \gls{NEP}, or building the structure of the virion by \gls{HA}, \gls{M1} and \gls{NA} \autocite{influenza_all}. % \DD{der Satz ist etwas unverstaendlich, evtl 2 Saetze draus machen?}. 
For a more effective way of replication, the genome of the \textit{Orthomyxoviridae} consists of not only one very long negative \gls{ssRNA} strand but instead between seven or eight smaller ones. The number of these small genomes, called segments, depends on the species %\DD{entweder 2 Saetze, Kommas rein damit das eindeutig ein Nebensatz ist oder Klammern drumrum} 
\autocite{influenza_all}. Each segment is coding for at least one of the major proteins needed for the life cycle of the virus \autocite{influenza_all}. Subtypes of the species \gls{IAV} from the \textit{Orthomyxoviridae} are named after variations of the surface proteins \gls{HA} and \gls{NA}, e.g. H5N1 or H1N1 \autocite{naming}. Only by connecting a receptor of the host cell with a matching antibody structure of the \gls{HA} protein on the virion surface, a cell can become infected, and only by matching \gls{NA} antibody, a new build virion can leave the host cell \autocite{influenza_all}.

Changes in the composition of the virion, especially related to the proteins on the surface can lead to drastic changes in the infectivity and %\DD{infectivity?} 
may result %\DD{"may result"} 
in even more dangerous pathogenic behavior \autocite{influenza_all}. These changes can occur by random mutations in the genome or by a process called reassortment. Mutation rates in negative \gls{ssRNA} viruses like \gls{IAV} are very high. One or two \glspl{SNP} occur in every virion from the quasispecies \autocite{mutation}. \glspl{SNP} are the change of one specific nucleotide to another by e.g. missing error control of the RNA polymerase complex. When a \gls{SNP} results in the change of the \gls{HA} or \gls{PA} protein structure, antigenic drift occurres %\DD{occurres} 
\autocite{drift}. The second and more drastic change in the virion structure of segmented viruses, like the \gls{IAV}, can happen by simultaneous infection of one host cell by more than one virion. The new replicated segments can be mixed together in the new virions potentially creating more dangerous combinations from viruses with different surface proteins and potentially enabling zoonosis. This process is called reassortment and leads to events called antigenic shift \autocite{shift}. Very dangerous, but human non-transferable subtypes, simultaneous infecting a non-human host cell with a human transferable, but harmless subtype, can reassort into a human transferable and dangerous subtypes \autocite{shift}.

Species from the family \textit{Orthomyxoviridae}, especially the \gls{IAV} and the \gls{IBV}, are a major threat to the human race by potentially developing deadly pandemics like the Spanish flu in 1918 \autocite{shift}. To prevent the composition of new deadly strains or weakening existing ones, better vaccines need to be developed. One way is to get a better knowledge about the secondary structure of the viruses, which is the way nucleotides bind to each other to accomplish various functions, %\DD{"things" klingt komisch, schreibs lieber um und nimm irgendwas wie "have various functions"} 
like starting the protein translation \autocite{ires}. With the mentioned virion structure and ways of infection in mind, the molecular mechanisms in the translation are of greater interest to find possible weak spots in the formation of vital structure proteins like \gls{HA} and \gls{NA} \autocite{influenza_all}. 

To synthesize the necessary proteins, each genomic negative \gls{ssRNA} segment is transcribed into positive \gls{mRNA}. The single \glspl{mRNA} can be divided into three pieces, the 5'-\gls{UTR}, \gls{CDS} and 3-\gls{UTR} \autocite{coding}. The \gls{CDS} contains the information to be translated into \glspl{AA} by ribosomes in form of three nucleotide long codons \autocite{buch}. The area of the \gls{CDS} starting with the start codon ATG and ending with a stop codon, usually AUG, is called \gls{ORF} and synthesizes a chain of \glspl{AA} later folded into a protein. Surrounding the \gls{CDS} on both sides are the 3'-\gls{UTR} on the 3'-end of the strand and the 5'-\gls{UTR} on the other side \autocite{buch}. \glspl{UTR} are non-coding sequences and usually not translated into peptides. Instead they can form important secondary structures to enable translation mechanisms, e.g. forming an \gls{IRES} which is in many viruses necessary for translation \autocite{ires}. 

Considering the high mutation rate of \textit{Orthomyxoviridae} and possible mechanisms to change its structure by reassembling its genome, it is not enough to get the secondary structure and weak spots of one quasispecies alone. To find wide-ranging vaccines, it is necessary to cluster complete strains of species and their various quasispecies into groups, find representatives of these groups and generate their secondary structure all together. That way weak spots in recurring and hopefully vital genomic regions can be discovered \autocite{cluster}. These groups are called clusters. For clustering various methods can be used. Clusters can be build by concepts, like trying to accomplish the best compactness of the sequences or using hierarchical approaches for connection \autocite{cluster}. Clustering techniques based on compactness ranging from comparison based on the nucleotide sequence identity threshold, like CD-HIT or usearch, to approaches using k-means algorithm \autocite{cluster, CD-HIT, Usearch, kmeans}. Clustering based on hierarchical clustering, like HDBSCAN connect sequences to cluster trees and use a given threshold to decide which sequences belongs to the same cluster \autocite{hdbscan}. 

The state of the art clustering tools usearch and CD-HIT are compared to a new tool named VeGETA using HDBSCAN for its clustering function \autocite{vegeta, Usearch, CD-HIT, hdbscan}. They were used on random subsets of various FASTA files containing the (segmented) genomes of all strains of \gls{IBV}. Clustering \gls{IAV} segments is skipped here, because the enormous number of highly different sequences for every segment makes random subsets extremely various. Some random sequences out of many ten thousands with highly different sequence, result in nearly the same number of clusters as used random sequences, especially with sequence identity based algorithms, like the ones used by usearch and CD-HIT. \gls{ICV} is skipped because there exist roughly 100 sequences per segment and the differences in these sequences are negligible, compared in a \gls{msa}. This is resulting in a small amount of clusters per tool, mostly only one, which is obstructive for the creation of a rating pipeline and meaningful statements about the clustering quality. The clusters of \gls{IBV} were compared and their quality rated by a calculated score, using the clusters representative, with a vector based rating method created in this project. After calculation of the used algorithms clustering quality, VeGETA is used on the same random subset of sequences, to calculate the overall secondary structure of the clustered sequences. These resulting secondary structures are validated by literature. 